File: 15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from typing import List

# @lc code=start
class Solution:
    # T: O(n ^ 2)
    # S: O(n) timsort
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        TARGET = 0
        n = len(nums)
        if n < 3:
            return []

        nums.sort()     # O(nlog(n)), timsort

        res = []
        for i in range(n - 2):
            x = nums[i]
            if x > TARGET:   
                # in a sorted list, sum of any 3 nums to the right of x is always greater than target
                return res
            if i > 0 and x == nums[i - 1]:
                continue    # avoid duplicates
            
            # Below is the same as Two Sum II Q167
            left, right = i + 1, n - 1
            while left < right:
                s = x + nums[left] + nums[right]
                if s == TARGET:    # == 0
                    res.append([x, nums[left], nums[right]])

                    # avoid duplicates
                    while left < right and nums[left] == nums[left+1]:
                        left += 1   
                    while left < right and nums[right] == nums[right-1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif s > TARGET:
                    right -= 1
                else:
                    left += 1
        return res

# @lc code=end

File: 15.三数之和.java
import java.util.*;

/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        if (n < 3)  return res;
        Arrays.sort(nums);  // O(N*log(N))

        for (int i = 0; i < n - 2; ++i) {
            int x = nums[i];
            if (x > 0)  return res; // after sorting, sum of any 3 nums to the right of x always > 0
            if (i > 0 && nums[i - 1] == x)  continue;   // skip duplicates
            int l = i + 1, r = n - 1;
            while (l < r) {
                int y = nums[l], z = nums[r];
                int sum = x + y + z;
                if (sum == 0) {
                    res.add(new ArrayList<>(Arrays.asList(x, y, z)));
                    while (l < r && nums[l + 1] == nums[l])  ++l;
                    while (l < r && nums[r - 1] == nums[r])  --r;
                    ++l;
                    --r;
                } else if (sum > 0) {
                    --r;
                } else {
                    ++l;
                }
            }
        }
        return res;
    }
}
// @lc code=end

